[Private sites] Redirect private atomic users to wp admin classic editor (from modal)

adamziel opened the pull request:
#### Changes proposed in this Pull Request

**Problem**

Classic editor in calypso does not support displaying private media files because of CORS issues. Adding a support would be super complex and probably not worth it, especially considering that we're phasing out the classic editor altogether. Classic editor in wp-admin handles private files out of the box so we're redirecting all private+atomic+classic editor users from calypso to wp-admin with `?classic_editor` attached to URL.

**Proposed solution**

We do have some dialog windows in place when user uses classic editor in Calypso:

Gutenberg content warning:

<img width="1090" alt="Zrzut ekranu 2020-03-13 o 16 32 06" src="https://user-images.githubusercontent.com/205419/76635769-3cc7d000-6548-11ea-8e48-ddbed3e6d744.png">

Gutenberg opt-in:

<img width="963" alt="Zrzut ekranu 2020-03-17 o 16 02 19" src="https://user-images.githubusercontent.com/205419/76869487-b291ab80-6868-11ea-965b-2156ecba5f6d.png">

My proposal is that for v1 we:
* Ensure that private+atomic user always sees one of these two modals when visiting classic editor. If no Gutenberg blocks are found in post content, let's display opt-in dialog.
* Ensure that choosing Classic editor redirects to corresponding editor in WP admin (ensured by 412-gh-wpcomsh).
* Make sure nothing changes for users who are not private+atomic+classic editor

#### Testing instructions

* Create a new private atomic site using a8c account and `calypso.localhost:3000` (start with simple site then go atomic - it should stay private)
* Apply related wpcomsh change from this PR **to the site you just created** 412-gh-wpcomsh
* Create a Gutenberg-based post, save it, in three dots menu select "Switch to classic editor"
* You should see Gutenberg content warning modal as on the screenshot above. Confirm that classic editor redirects to classic editor in wp-admin.
* Go back to calypso, create new page
* You should see Gutenberg opt-in modal as on the screenshot above. Confirm that choosing classic editor redirects to classic editor in wp-admin.
* Try on a non-private atomic site and confirm the behavior is the same as currently in production, e.g. you're not getting redirected to `wp-admin` and you're not getting opt-in modal every single time

p-jackson commented on client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx:
```diff
@@ -78,9 +94,29 @@ class EditorGutenbergBlocksWarningDialog extends Component {
+
+	redirectToWpAdmin = () => {
+		const { isAtomic, isPrivate, postId, postType, buildSiteAdminUrl } = this.props;
+		if ( isAtomic && isPrivate ) {
```
I've only had a quick look at this, but if this is `false` should we be calling `setState( { forceClassic: true } )`?

I'm thinking of whether we need to retain the existing `useClassicEditor()` behaviour. Although we wouldn't want to call `setState( { forceClassic: true } )` if `redirectToWpAdmin()` was being called from `cDM` 🤔

adamziel replied:
Good note! I tweaked it slightly

p-jackson commented:
Issue when using the classic editor in `wp-admin`. After landing in the old classic editor and saving a draft, the page reloads, but it loads the block editor.
See video: https://d.pr/v/lRMxwz

Could it be that clicking **Save Draft** is dropping the `?classic_editor` parameter? Feels like there could be lots of places in the UI that reload the page without forwarding the param.

p-jackson commented:
After switching to the classic editor in the above comment, the classic editor seems to become my preferred editor. But when I open the editor from calypso it opens the classic editor in calypso and doesn't redirect to `wp-admin`.
See video: https://d.pr/v/SljGs1

Maybe we can't rely on putting the `?classic_editor` param in all the right places. Could `wpcomsh` use atomic and privacy settings to determine where to open the editor rather than a parameter?

adamziel commented:
> Issue when using the classic editor in wp-admin. After landing in the old classic editor and saving a draft, the page reloads, but it loads the block editor.

Alright, it's because it was only redirecting after clicking the button. It didn't account for the fact that the modal may be not displayed at all. I just fixed that.

> Maybe we can't rely on putting the ?classic_editor param in all the right places. Could wpcomsh use atomic and privacy settings to determine where to open the editor rather than a parameter?

Fortunately this is accounted for in the [classic editor plugin](https://plugins.trac.wordpress.org/browser/classic-editor/trunk/classic-editor.php#L595) - it didn't happen because of a subtle issue in related wpcomsh PR. I just fixed it so should be all good now. 🤞

noahtallen commented on client/post-editor/editor-gutenberg-opt-in-dialog/index.jsx:
```diff
@@ -42,10 +45,16 @@ class EditorGutenbergOptInDialog extends Component {
-		this.props.hideDialog();
+		const { isPrivateAtomic, wpAdminRedirectionUrl } = this.props;
+		if ( isPrivateAtomic ) {
+			window.location.href = wpAdminRedirectionUrl;
```
Is there a scenario where this path could execute even if we're trying to use Gutenberg?

adamziel replied:
I just renamed things a little bit to make it more obvious what's going on

p-jackson commented on client/state/selectors/get-wp-admin-classic-editor-redirection-url.ts:
```diff
@@ -0,0 +1,39 @@
+		action: postId && 'edit', // If postId is set, open edit view.
+		post_type: postType !== 'post' && postType, // Use postType if it's different than post.
+		'classic-editor': 1,
+	} );
```
TIL this cool trick of using `pickBy()` to remove the falsey ones :)

p-jackson commented on client/post-editor/editor-gutenberg-dialogs/index.tsx:
```diff
@@ -0,0 +1,59 @@
+				dispatch( showGutenbergOptInDialog() );
+			}
+		},
+		[ isPrivateAtomic, isPostContentLoaded, postContent ]
```
Dependency array is missing `dispatch` and `hasGutenbergBlocks`

adamziel replied:
> React guarantees that dispatch function identity is stable and won't change on re-renders. This is why it's safe to omit from the useEffect or useCallback dependency list.
https://reactjs.org/docs/hooks-reference.html#usereducer

adamziel replied:
Regarding `hasGutenbergBlocks`, it's only set in that effect and it's meant to be only set once (if it's currently `null`) so it's missing on purpose - we don't want to rerun this effect after it sets `hasGutenbergBlocks`

p-jackson replied:
True. We're using the `dispatch` from `react-redux`, but it makes sense that they would make the same guarantee. Although this example from their docs includes it in the dependency list 🤔 
https://react-redux.js.org/next/api/hooks#examples

What initially alerted me to the issue was eslint, it's up to you whether you think it's better to just add the dependencies or add an eslint skip comment

adamziel replied:
Yeah this should really be documented. Here's an info I found in a [similar discussion in another repo](https://github.com/reduxjs/redux-toolkit/issues/293):

> Usually, you do not need to, as dispatch will not change unless the react-redux Prodvider gets a new store and in a normal application that should never be the acse.
> But generally, it is considered best practice to just use the react eslint rules, and those do warn you if you don't add it in the dependencies array. So I'd just add it there, as it has no negative impact and you will follow best practices.

I'll just add it to dependencies

adamziel replied:
And then I also had to add `// eslint-disable-next-line react-hooks/exhaustive-deps` to make lint OK with `hasGutenbergBlocks` missing

simison replied:
>had to add // eslint-disable-next-line react-hooks/exhaustive-deps to make the linter OK with having hasGutenbergBlocks missing

Would be good to add note why it was disabled in case it's not obvious.

adamziel commented on client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx:
```diff
@@ -33,19 +36,21 @@ import './style.scss';
 		translate: PropTypes.func,
-		postContent: PropTypes.string,
 		siteId: PropTypes.number,
+		postType: PropTypes.string,
```
Done!

adamziel commented:
Thank you for thorough feedback @noahtallen! I also think we should just move forward here even if it's not bullet-proof. This is a big PR that shouldn't affect any users for now. It's slow to iterate on so let's get it merged and discuss any fixes in atomic-ish PRs.

> I really think we should just be redirected immediately after hitting the /post route if the user has opted into the classic editor

This was my first idea and I even started a PR that does just that (#40123). There is one huge problem with that approach - there is no way to switch to Gutenberg in Calypso. You can still switch to Gutenberg in wp-admin, but as you noticed it does not affect the preferred editor in Calypso. Displaying the modal is not perfect either, but at least it provides you a way to change your mind and migrate to Gutenberg.

What should really happen, is we should take two steps back, involve designers, and rethink the entire experience of Classic editor vs Gutenberg. This is definitely out of scope for this project, hence the modal workaround. I believe it will be a part of p58i-8Hc-p2 though

> I actually want to highlight this a bit more. It seems like the modal doesn't always show up. I can't figure out the conditions in which the modal does not appear, but it has remained hidden at least once or twice while I've been testing.

I found one possible answer in `lib/gutenberg/editing-flow.php#79` which disables Gutenberg opt-in modals in case classic editor plugin is active, and disables both modals in turn. I think in this case we should just redirect to wp-admin. I may add it to this PR and merge, or spin a tiny follow-up PR.

adamziel commented:
@noahtallen the change to support no modals case turned out to be small, but warranting it's own PR: https://github.com/Automattic/wp-calypso/pull/40362

noahtallen commented:
> rethink the entire experience of Classic editor vs Gutenberg... I believe it will be a part of p58i-8Hc-p2 though

Yeah, sounds good to me!
